### PR TITLE
Coordinates widget

### PR DESCRIPTION
**Pull request by nnngrach**
https://github.com/osmandapp/OsmAnd-Issues/issues/424

video:
https://www.dropbox.com/s/rwz1rw1igv3ydou/screen%20recording%202021-03-30%20at%2004.46.36.mov?dl=0

https://www.dropbox.com/s/ckvb7s3icto48e3/screen%20recording%202021-03-30%20at%2004.49.03.mov?dl=0

### CODE REVIEW

File: Resources/en.lproj/Localizable.strings [hunk 1]
```diff
@@ -1153,13 +1154,22 @@
...
 "quick_action_name" = "Quick action";
 "quick_action_no_overlay" = "No overlay";
 "quick_action_add_overlay" = "Add overlay";
 "quick_action_no_underlay" = "No underlay";
 "quick_action_add_underlay" = "Add underlay";
 "quick_action_list_descr" = "Tapping this action button pages through the list below";
+"toggle_coordinates" = "Show/hide coordinates widget";
+"show_coordinates" = "Show coordinates";
+"hide_coordinates" = "Hide coordinates";
+"quick_action_coordinates_widget_descr" = "A toggle to show or hide the Coordinates widget on the map.";
```
Comment (PaulStets) [inline Resources/en.lproj/Localizable.strings:1172]: Description should be written in the same style as existing actions.
Example:
"quick_action_hillshade_descr" = "Tapping this action button shows or hides the terrain layer.";

File: Sources/Controllers/Map/OADestinationViewController.mm [hunk 2]
```diff
@@ -419,6 +419,7 @@ - (void)updateFrame:(BOOL)animated
                 _multiCell.contentView.hidden = YES;
 
             for (OADestinationCell *cell in _destinationCells)
+            {
```
Comment (PaulStets) [inline Sources/Controllers/Map/OADestinationViewController.mm:422]: Will this compile?

File: Sources/Controllers/Map/OAMapHudViewController.mm [hunk 3]
```diff
@@ -790,6 +813,28 @@ - (void) updateButtonsLayoutY:(CGFloat)y
...
+    BOOL isCoordinatesVisible = [OAAppSettings.sharedManager.showCoordinatesWidget get];
+    BOOL isMarkersWidgetVisible = yOffset > 0;
+    CGFloat markersWidgetHeaderHeight = 50;
+    CGFloat coordinateWidgetHeight = 52;
+    if ([OAUtilities isLandscape])
+    {
+        if (isCoordinatesVisible && isMarkersWidgetVisible)
+            return 0;
+        else if (isMarkersWidgetVisible)
+            return -1 * markersWidgetHeaderHeight;
```
Comment (PaulStets) [inline Sources/Controllers/Map/OAMapHudViewController.mm:827]: Simply adding - is enough

File: Sources/Helpers/OAAppSettings.m [hunk 4]
```diff
@@ -2020,11 +2021,13 @@ - (instancetype) init
         _showIntermediateArrivalTime = [OAProfileBoolean withKey:showIntermediateArrivalTimeKey defValue:YES];
         _showRelativeBearing = [OAProfileBoolean withKey:showRelativeBearingKey defValue:YES];
         _showCompassControlRuler = [OAProfileBoolean withKey:showCompassControlRulerKey defValue:YES];
+        _showCoordinatesWidget = [OAProfileBoolean withKey:showCoordinatesWidgetKey defValue:YES];
```
Comment (PaulStets) [inline Sources/Helpers/OAAppSettings.m:2024]: Default value is false in android.

File: Sources/Controllers/Map/OAMapInfoController.mm [hunk 5]
```diff
@@ -408,6 +411,25 @@ - (void) layoutWidgets:(OATextInfoWidget *)widget
         CGRect f = _rightWidgetsView.superview.frame;
         _rightWidgetsView.superview.frame = CGRectMake(f.origin.x, f.origin.y, f.size.width, maxContainerHeight);
     }
+    
+    if (_topCoordinatesView && _topCoordinatesView.superview && !_topCoordinatesView.hidden)
+    {
+        if (_lastUpdateTime == 0)
+            [[OARootViewController instance].mapPanel updateToolbar];
+        
+        BOOL isMarkerVidgetVisible = [_mapHudViewController getControlsTopPosition] > 0;
```
Comment (alex-osm) [inline Sources/Controllers/Map/OAMapInfoController.mm:420]: Looks like wrong name. Is getControlsTopPosition related to Marker only?
Vidget > Widget

File: Sources/Controllers/Map/Widgets/OATopCoordinatesWidget.m [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+- (int) getConverterFormat:(int)format
+{
+    return format + 101;
+}
+
+- (BOOL) shouldUpdate
+{
+    BOOL isFirstLaunch = _lastUpdatingTime == 0;
+    
+    BOOL isVisibilityChabged = !isFirstLaunch && _cachedVisibiliy != [_settings.showCoordinatesWidget get];
```
Comment (alex-osm) [inline Sources/Controllers/Map/Widgets/OATopCoordinatesWidget.m:231]: isVisibilityChabged > isVisibilityChanged
isFirstLaunch is not related to visibility change.

File: Sources/Controllers/Map/Widgets/OATopCoordinatesWidget.m [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+    
+    CLLocation *currentLocation = _app.locationServices.lastKnownLocation;
+    BOOL locationNotChanged = [OAUtilities isCoordEqual:currentLocation.coordinate.latitude srcLon:currentLocation.coordinate.longitude destLat:_lastKnownLocation.coordinate.latitude destLon:_lastKnownLocation.coordinate.latitude];
+    
+    NSTimeInterval updatingPeriond = 0.5;
+    NSTimeInterval currentTimestamp = [[NSDate new] timeIntervalSince1970];
+    NSTimeInterval difference = currentTimestamp - _lastUpdatingTime;
+    BOOL notEnoughTimePassed = currentTimestamp - _lastUpdatingTime < updatingPeriond;
+    
+    return isFirstLaunch || isVisibilityChabged || (!_isAnimated && !locationNotChanged && !notEnoughTimePassed);
```
Comment (alex-osm) [inline Sources/Controllers/Map/Widgets/OATopCoordinatesWidget.m:241]: Do not write "!not". locationNotChanged > locationChanged, etc.

File: Sources/Helpers/OADiscountHelper.h [hunk 8]
```diff
@@ -53,5 +53,6 @@
 + (OADiscountHelper *)instance;
 
 - (void) checkAndDisplay;
+- (BOOL) isVisible;
```
Comment (alex-osm) [inline Sources/Helpers/OADiscountHelper.h:56]: Drop it. DiscountBanner is a ToolbarViewController. It seems you should check for any toolbar visibility.

**Pull request merged**